### DISCUSSION TITLE
check_nrpe using NSClient++ gives random socket timeout

### DISCUSSION THREAD

**Discussion by futureweb** [mickem/nscp#829] [Q&A]
Hey there,
already roamed through dozens of postings here - but unfortunately none of them helped me solving this problem. Maybe one of you got another idea which helps me?! :-)

I'm using check_nrpe to NSClient++ for Monitoring some Windows Servers.
"Normal" Checks (Disk/CPU/Memory/...) all working fine and reliable, but the "Check HyperV Replication Status" Check is flapping massively. About every other check it gets "Socket Timeout" ... Already tried different Versions of check_nrpe - example:

```
#  /dl/nrpe-4.1.0/src/check_nrpe  -H server.dom.tld -t 60 -u -p 5668 -c CheckHyperVReplica  -2 -4
Normal
#  /dl/nrpe-4.1.0/src/check_nrpe  -H server.dom.tld-t 60 -u -p 5668 -c CheckHyperVReplica  -2 -4
Normal
#  /dl/nrpe-4.1.0/src/check_nrpe  -H server.dom.tld -t 60 -u -p 5668 -c CheckHyperVReplica  -2 -4
CHECK_NRPE STATE UNKNOWN: Socket timeout after 60 seconds.
#  /dl/nrpe-4.1.0/src/check_nrpe  -H server.dom.tld -t 60 -u -p 5668 -c CheckHyperVReplica  -2 -4
Normal
#  /dl/nrpe-4.1.0/src/check_nrpe  -H server.dom.tld -t 60 -u -p 5668 -c CheckHyperVReplica  -2 -4
CHECK_NRPE STATE UNKNOWN: Socket timeout after 60 seconds.
```

So it's completely random when it fails and when it succeeds. When executing the Check Hyper-V Replica PS Script on Server it always responds in time (5-15 seconds) and never failed / timeout.

Anyone any idea on this? :-)

Thank you
Andy

**Comment by mickem** [comment]
Sorry for the late reply. Since the built-in checks are stable and only the PowerShell one flaps, even when you loop `check_nrpe` by hand, I would not look at the network or at check_nrpe. The pattern matches a well-known problem with `powershell.exe` started from a service: with a redirected console it sometimes blocks waiting on stdin (or on a console handle) and never exits, so NSClient++ waits on the process until check_nrpe gives up. Which run hangs is timing dependent, which is why it looks random.

Three things to check:

1. **How is the script launched?** Post the `[/settings/external scripts/...]` lines for `CheckHyperVReplica`. If it is a plain `powershell.exe -File ...` line, add `-NonInteractive -NoProfile -ExecutionPolicy Bypass -InputFormat None`:

   ```ini
   [/settings/external scripts/scripts]
   CheckHyperVReplica = powershell.exe -NoProfile -NonInteractive -ExecutionPolicy Bypass -InputFormat None -File "scripts\CheckHyperVReplica.ps1"
   ```

   If it is a *wrapped* script, the same flags go into the `ps1` wrapping under `[/settings/external scripts/wrappings]`. Current releases ship that as the default (`powershell.exe -NoProfile -ExecutionPolicy Bypass -NonInteractive -File "scripts\%SCRIPT%" ...`), so upgrading to a current release fixes it for wrapped scripts on its own.

2. **What does the agent see?** Set `[/settings/log] level = debug`, restart the service, reproduce a timeout and look at `nsclient.log` around it. You will either see the script start and never finish (the hang above), or finish late, or see the connection arrive after something else. Each of those points somewhere different, so please paste that section.

3. **Timeouts.** `[/settings/external scripts] timeout` defaults to 60 seconds, the same as your `-t 60`, so when the script hangs the agent's own kill and check_nrpe's timeout race and you get the socket timeout rather than a clean "script timed out" message. Setting the agent's timeout below check_nrpe's (say 45) makes a hang show up as a proper UNKNOWN with a message instead of a socket error, which also makes the log easier to read.

If after the flags it still hangs, the next suspect is the Hyper-V cmdlets themselves under the service account (`Get-VMReplication`/`Measure-VMReplication` go through the VMMS WMI provider, which can stall); the debug log timing will tell us which.
